File: sdk/typescript/cbor.mjs
const concat = (parts) => Buffer.concat(parts.map((part) => Buffer.from(part)));

function head(major, value) {
  if (value < 24) return Buffer.from([(major << 5) | value]);
  if (value <= 0xff) return Buffer.from([(major << 5) | 24, value]);
  if (value <= 0xffff) {
    const result = Buffer.alloc(3); result[0] = (major << 5) | 25; result.writeUInt16BE(value, 1); return result;
  }
  if (value <= 0xffffffff) {
    const result = Buffer.alloc(5); result[0] = (major << 5) | 26; result.writeUInt32BE(value, 1); return result;
  }
  const result = Buffer.alloc(9); result[0] = (major << 5) | 27;
  result.writeBigUInt64BE(BigInt(value), 1); return result;
}

export function encode(value) {
  if (value === null || value === undefined) return Buffer.from([0xf6]);
  if (value === false) return Buffer.from([0xf4]);
  if (value === true) return Buffer.from([0xf5]);
  if (typeof value === "number" && Number.isSafeInteger(value))
    return value >= 0 ? head(0, value) : head(1, -1 - value);
  if (typeof value === "number") {
    const result = Buffer.alloc(9); result[0] = 0xfb; result.writeDoubleBE(value, 1); return result;
  }
  if (typeof value === "string") {
    const bytes = Buffer.from(value, "utf8"); return concat([head(3, bytes.length), bytes]);
  }
  if (Buffer.isBuffer(value) || value instanceof Uint8Array) {
    const bytes = Buffer.from(value); return concat([head(2, bytes.length), bytes]);
  }
  if (Array.isArray(value)) return concat([head(4, value.length), ...value.map(encode)]);
  if (typeof value === "object") {
    const entries = Object.entries(value).sort(([a], [b]) =>
      Buffer.byteLength(a) - Buffer.byteLength(b) || Buffer.from(a).compare(Buffer.from(b)));
    return concat([head(5, entries.length), ...entries.flatMap(([key, item]) => [encode(key), encode(item)])]);
  }
  throw new TypeError(`unsupported CBOR value: ${typeof value}`);
}

export function decode(buffer) {
  const bytes = Buffer.from(buffer); let offset = 0;
  const argument = (additional) => {
    if (additional < 24) return additional;
    if (additional === 24) return bytes[offset++];
    if (additional === 25) { const value = bytes.readUInt16BE(offset); offset += 2; return value; }
    if (additional === 26) { const value = bytes.readUInt32BE(offset); offset += 4; return value; }
    if (additional === 27) { const value = Number(bytes.readBigUInt64BE(offset)); offset += 8; return value; }
    throw new Error("indefinite CBOR is not supported");
  };
  const read = (depth = 0) => {
    if (depth > 64 || offset >= bytes.length) throw new Error("invalid CBOR frame");
    const initial = bytes[offset++]; const major = initial >> 5; const additional = initial & 31;
    if (major === 7) {
      if (additional === 20) return false; if (additional === 21) return true;
      if (additional === 22) return null;
      if (additional === 27) { const value = bytes.readDoubleBE(offset); offset += 8; return value; }
      throw new Error("unsupported CBOR simple value");
    }
    const length = argument(additional);
    if (major === 0) return length; if (major === 1) return -1 - length;
    if (major === 2) { const value = bytes.subarray(offset, offset + length); offset += length; return value; }
    if (major === 3) { const value = bytes.toString("utf8", offset, offset + length); offset += length; return value; }
    if (major === 4) return Array.from({ length }, () => read(depth + 1));
    if (major === 5) {
      const value = {}; for (let index = 0; index < length; index += 1) value[read(depth + 1)] = read(depth + 1);
      return value;
    }
    throw new Error("unsupported CBOR major type");
  };
  const value = read(); if (offset !== bytes.length) throw new Error("trailing CBOR bytes"); return value;
}

File: sdk/python/tokmon_lens_sdk/cbor.py
from __future__ import annotations

import struct
from typing import Any


def _head(major: int, value: int) -> bytes:
    prefix = major << 5
    if value < 24:
        return bytes([prefix | value])
    if value <= 0xFF:
        return bytes([prefix | 24, value])
    if value <= 0xFFFF:
        return bytes([prefix | 25]) + struct.pack(">H", value)
    if value <= 0xFFFFFFFF:
        return bytes([prefix | 26]) + struct.pack(">I", value)
    return bytes([prefix | 27]) + struct.pack(">Q", value)


def encode(value: Any) -> bytes:
    if value is None:
        return b"\xf6"
    if value is False:
        return b"\xf4"
    if value is True:
        return b"\xf5"
    if isinstance(value, int):
        return _head(0, value) if value >= 0 else _head(1, -1 - value)
    if isinstance(value, float):
        return b"\xfb" + struct.pack(">d", value)
    if isinstance(value, str):
        data = value.encode("utf-8")
        return _head(3, len(data)) + data
    if isinstance(value, (bytes, bytearray, memoryview)):
        data = bytes(value)
        return _head(2, len(data)) + data
    if isinstance(value, (list, tuple)):
        return _head(4, len(value)) + b"".join(encode(item) for item in value)
    if isinstance(value, dict):
        items = sorted(value.items(), key=lambda item: (len(item[0].encode()), item[0].encode()))
        return _head(5, len(items)) + b"".join(
            encode(str(key)) + encode(item) for key, item in items
        )
    raise TypeError(f"unsupported CBOR value: {type(value)!r}")


def decode(data: bytes) -> Any:
    offset = 0

    def argument(additional: int) -> int:
        nonlocal offset
        if additional < 24:
            return additional
        sizes = {24: 1, 25: 2, 26: 4, 27: 8}
        size = sizes.get(additional)
        if size is None or offset + size > len(data):
            raise ValueError("invalid CBOR argument")
        value = int.from_bytes(data[offset:offset + size], "big")
        offset += size
        return value

    def read(depth: int = 0) -> Any:
        nonlocal offset
        if depth > 64 or offset >= len(data):
            raise ValueError("invalid CBOR frame")
        initial = data[offset]
        offset += 1
        major, additional = initial >> 5, initial & 31
        if major == 7:
            if additional == 20:
                return False
            if additional == 21:
                return True
            if additional == 22:
                return None
            if additional == 27:
                value = struct.unpack_from(">d", data, offset)[0]
                offset += 8
                return value
            raise ValueError("unsupported CBOR simple value")
        length = argument(additional)
        if major == 0:
            return length
        if major == 1:
            return -1 - length
        if major == 2:
            value = data[offset:offset + length]
            offset += length
            return value
        if major == 3:
            value = data[offset:offset + length].decode("utf-8")
            offset += length
            return value
        if major == 4:
            return [read(depth + 1) for _ in range(length)]
        if major == 5:
            return {read(depth + 1): read(depth + 1) for _ in range(length)}
        raise ValueError("unsupported CBOR major type")

    value = read()
    if offset != len(data):
        raise ValueError("trailing CBOR bytes")
    return value

